### PR TITLE
*: Chunkreuse try reuse chunk in operator

### LINKED ISSUE #38606: chunk: reuse tidb chunk
## Enhancement
Reuse chunks and columns in conn .

### PR DESCRIPTION
**Pull request by keeplearning20221**
### What problem does this PR solve?

Issue Number: close #38606

Problem Summary:
Every time chunk is used, it will be applied from the system, which affects the execution efficiency of tidb-server.

### What is changed and how it works?
Cache previously applied chunks and columns in alloc.later, it will get chunk from cache first , if not enough then apply from system.
Alloc modified based on the alloc developed by tiancaiamao. It uses slices to manage chunks objects and columns  objects. When used, they are taken out of free slices and placed in allocated slices.When the SQL execution is complete, put it back from the allocated slice back to the free slice.
Add configuration items
tidb_enable_reuse_chunk     function switch
tidb_max_reuse_chunk         Maximum number of cached chunks
tidb_max_reuse_column       Maximum number of cached columns
Add read-only variable：
Last_sql_use_alloc          Whether the previous sql uses chunk alloc
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
@keeplearning20221 Please fix the linter.

**Comment by hawkingrei** [comment]
@keeplearning20221 You can try to run TPCC by benchbot to verify your work.

File: server/conn_stmt.go [hunk 1]
```diff
@@ -295,6 +299,7 @@ const (
 
 func (cc *clientConn) handleStmtFetch(ctx context.Context, data []byte) (err error) {
 	cc.ctx.GetSessionVars().StartTime = time.Now()
+	cc.ctx.GetSessionVars().EndAlloc()
```
Comment (tiancaiamao) [inline server/conn_stmt.go:302]: Fetch is used by the prepared/exec protocol?

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -1281,6 +1281,68 @@ type SessionVars struct {
...
+	// ChunkPool Several chunks and columns are cached
+	ChunkPool struct {
+		Lock  sync.Mutex
+		Alloc chunk.Allocator
+	}
+	EnableReuseCheck bool
+
+	// UseChunkAlloc indicates whether statement use chunk alloc
+	UseChunkAlloc    bool
+	PreUseChunkAlloc bool
```
Comment (tiancaiamao) [inline sessionctx/variable/session.go:1296]: Will the outside package change those field? if not, you can use lowercase like useChunkAlloc / perUseChunkAlloc etc

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -1281,6 +1281,68 @@ type SessionVars struct {
...
+// GetNewChunkWithCapacity Attempt to request memory from the chunk pool
+// thread safety
+func (s *SessionVars) GetNewChunkWithCapacity(fields []*types.FieldType, capacity int, maxCachesize int) *chunk.Chunk {
+	if s.ChunkPool.Alloc == nil {
+		return chunk.New(fields, capacity, maxCachesize)
+	}
+	s.ChunkPool.Lock.Lock()
+	defer s.ChunkPool.Lock.Unlock()
+	if s.MaxReuseChunk > 0 || s.MaxReuseColumn > 0 && (!s.UseChunkAlloc) {
+		s.UseChunkAlloc = true
```
Comment (tiancaiamao) [inline sessionctx/variable/session.go:1324]: Is the variable necessary? Maybe you can defile a function?

```
func (s *SessionVars) UseChunkAlloc() bool {
    return s.MaxReuseChunk > 0 ||  s.MaxReuseColumn > 0
}
```

The different is that when you have several variable for something, and the value of some variables are derived from the others, it's easy to get into consistency between those variables.

File: server/conn.go [hunk 4]
```diff
@@ -1105,6 +1105,7 @@ func (cc *clientConn) Run(ctx context.Context) {
 		startTime := time.Now()
 		err = cc.dispatch(ctx, data)
 		cc.chunkAlloc.Reset()
+		cc.ctx.GetSessionVars().EndAlloc()
```
Comment (tiancaiamao) [inline server/conn.go:1108]: `EndAlloc` every query? so how is the object reused?

File: server/conn_stmt.go (hunk 1 above)
Comment (keeplearning20221) [reply server/conn_stmt.go:302]: used by fetch request（Maybe java api or c api）

File: sessionctx/variable/session.go (hunk 2 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go:1296]: UT test cases check the state of these variables. Need  change to function ?

File: sessionctx/variable/session.go (hunk 3 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go:1324]: ok

File: server/conn.go (hunk 4 above)
Comment (keeplearning20221) [reply server/conn.go:1108]: The alloc object in the connection-level cc.chunkAlloc, and is referenced to the session-level pool every time it needs to be used.

**Review by tiancaiamao — APPROVED** [review]

File: sessionctx/variable/session.go (hunk 2 above)
Comment (tiancaiamao) [reply sessionctx/variable/session.go:1296]: It's up to you.
Use lowercase and provide visitor function sometimes can help avoid the variables been modified unexpectedly, but make the code longer

File: sessionctx/variable/session.go (hunk 3 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go]: done

File: sessionctx/variable/session.go (hunk 2 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go]: done

**Comment by bob34007** [comment]
LGTM

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -1281,6 +1281,84 @@ type SessionVars struct {
...
+// thread safety
+func (s *SessionVars) GetNewChunk(fields []*types.FieldType, capacity int) *chunk.Chunk {
+	//Chunk memory pool is not set
+	if s.ChunkPool.Alloc == nil {
+		return chunk.NewChunkWithCapacity(fields, capacity)
+	}
+	s.ChunkPool.Lock.Lock()
+	defer s.ChunkPool.Lock.Unlock()
+	if s.checkReuseAllocSize() && (!s.useChunkAlloc) {
+		s.useChunkAlloc = true
```
Comment (cfzjywxk) [inline sessionctx/variable/session.go:1309]: Seems the `useChunkAlloc` would be reset for each statement, maybe it's more appropriate to put it into the `stmtContext`. and use it in the later statement processing.

File: sessionctx/variable/sysvar.go [hunk 6]
```diff
@@ -1954,10 +1957,24 @@ var defaultSysVars = []*SysVar{
...
+			return nil
+		}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBMaxReuseChunk, Value: strconv.Itoa(DefTiDBMaxReuseChunk), Type: TypeUnsigned, MinValue: 1, MaxValue: math.MaxInt32, SetSession: func(s *SessionVars, val string) error {
+		s.MaxReuseChunk = tidbOptPositiveInt32(val, DefTiDBMaxReuseChunk)
+		return nil
+	}},
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBMaxReuseColumn, Value: strconv.Itoa(DefTiDBMaxReuseColumn), Type: TypeUnsigned, MinValue: 1, MaxValue: math.MaxInt32, SetSession: func(s *SessionVars, val string) error {
+		s.MaxReuseColumn = tidbOptPositiveInt32(val, DefTiDBMaxReuseColumn)
+		return nil
+	}},
```
Comment (cfzjywxk) [inline sessionctx/variable/sysvar.go:1977]: The newly introduced variables need to be approved by the PM. IMO, the user may not know how to configure these parameters.

File: sessionctx/variable/session.go (hunk 5 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go]: done

File: sessionctx/variable/session.go [hunk 7]
```diff
@@ -1281,6 +1281,84 @@ type SessionVars struct {
...
+	// ChunkPool Several chunks and columns are cached
+	ChunkPool struct {
+		Lock  sync.Mutex
+		Alloc chunk.Allocator
+	}
+	EnableReuseCheck bool
+
+	// useChunkAlloc indicates whether statement use chunk alloc
+	useChunkAlloc    bool
+	preUseChunkAlloc bool
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply sessionctx/variable/session.go]: done

File: util/chunk/chunk.go [hunk 8]
```diff
@@ -116,6 +116,15 @@ func renewColumns(oldCol []*Column, capacity int) []*Column {
 	return columns
 }
 
+// renewColumnsWithCache try renew Column from cache
+func renewColumnsWithCache(oldCol []*Column, capacity int, alloc allocator) []*Column {
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply util/chunk/chunk.go]: removed

File: executor/executor.go [hunk 9]
```diff
@@ -2171,6 +2177,10 @@ func ResetUpdateStmtCtx(sc *stmtctx.StatementContext, stmt *ast.UpdateStmt, vars
 // expression using rows from a chunk, and then fill this value into the chunk
 func FillVirtualColumnValue(virtualRetTypes []*types.FieldType, virtualColumnIndex []int,
 	schema *expression.Schema, columns []*model.ColumnInfo, sctx sessionctx.Context, req *chunk.Chunk) error {
+	if len(virtualColumnIndex) == 0 {
```
[the remark this answers is not in the document]
Comment (keeplearning20221) [reply executor/executor.go:2180]: In this way, reduce application of a chunk

File: sessionctx/variable/session.go [hunk 10]
```diff
@@ -1284,6 +1284,84 @@ type SessionVars struct {
...
+	ChunkPool struct {
+		Lock  sync.Mutex
+		Alloc chunk.Allocator
+	}
+	// EnableReuseCheck indicates  request chunk whether use chunk alloc
+	EnableReuseCheck bool
+
+	// preuseChunkAlloc indicates whether pre statement use chunk alloc
+	// like select @@last_sql_use_alloc
+	preUseChunkAlloc bool
```
Comment (cfzjywxk) [inline sessionctx/variable/session.go:1302]: Does it mean that the previous statement uses chunk buffer allocation? When would this variable be used, could you give some examples or situations?

**Comment by XuHuaiyu** [comment]
Any performance test and memory consumption test on this commit?

**Comment by XuHuaiyu** [comment]
1. Why change the default value of `maxFreeChunks  = 64` and `maxFreeColumnsPerType = 256`? It'll raise the OOM risk.
2. How much memory will the chunk pool use by default?

File: sessionctx/variable/session.go (hunk 10 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go:1308]: CREATE TABLE `t1` (
  `id` int(11) NOT NULL,
  `id2` int(11) DEFAULT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ;
insert into t1 values (1,1);
MySQL [test]> select * from t1 where id2 = 2;
+----+------+
| id | id2  |
+----+------+
|  2 |    2 |
+----+------+
1 row in set (0.00 sec)

MySQL [test]> select @@Last_sql_use_alloc;
+----------------------+
| @@Last_sql_use_alloc |
+----------------------+
|                    1 |
+----------------------+
1 row in set (0.00 sec)

MySQL [test]> select * from t1 where id = 1;
+----+------+
| id | id2  |
+----+------+
|  1 |    1 |
+----+------+
1 row in set (0.00 sec)

MySQL [test]> select @@Last_sql_use_alloc;
+----------------------+
| @@Last_sql_use_alloc |
+----------------------+
|                    0 |
+----------------------+
1 row in set (0.00 sec)

**Comment by keeplearning20221** [comment]
> Any performance test and memory consumption test on this commit?

https://pingcap.feishu.cn/docx/ZvgwdD6PpoKH0KxlQBCcjF1Nnsc

File: sessionctx/variable/session.go (hunk 10 above)
Comment (keeplearning20221) [reply sessionctx/variable/session.go:1308]: added ut test

**Review by cfzjywxk — APPROVED** [review]
Perhaps related regression tests and bechmark tests are needed as it would be enabled by default.

**Comment by tiancaiamao** [comment]
> Perhaps related regression tests and bechmark tests are needed as it would be enabled by default.

We can change to disable by default if we find anything wrong during test.

**Pull request merged**